### PR TITLE
Add one-to-one relation with raw tracker/calorimeter hit for reconstructed tracker/calorimeter hit

### PR DESCRIPTION
**Pull request by rahmans1**
### Briefly, what does this PR introduce?
Add one-to-one relation with raw tracker/calorimeter hit for reconstructed tracker/calorimeter hit. Contingent on https://github.com/eic/EDM4eic/pull/86. 

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [ ] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ ] Changes have been communicated to collaborators

### CODE REVIEW

File: src/algorithms/calorimetry/CalorimeterHitReco.cc [hunk 1]
```diff
@@ -276,6 +276,7 @@ void CalorimeterHitReco::process(
             sid,
             lid,
             local_position);
+        recohits->setRawHit(rh);
```
Comment (veprbl) [inline src/algorithms/calorimetry/CalorimeterHitReco.cc:279]: For the final version of this PR, we need to update those to support older EDM's.
```suggestion
#if EDM4EIC_VERSION_MAJOR >= 7
        recohits->setRawHit(rh);
#endif
```

and
```c++
#include <edm4eic/EDM4eicVersion.h>
```

**Comment by wdconinc** [comment]
I think there is a desire to move forward on tracking here, where the implementation is essentially complete thanks to one-to-one associations and relations, but I fear that if we add calorimeter hit associations and relations to the scope of this PR it will be slowed down. In particular, the CalorimeterHitDigi does pre-digitization merging of SimCalorimeterHits, so the associations are going to be requiring some non-trivial code.

So, I suggest we keep this PR for tracker (and photomultiplier) hits only, and create a similar PR for calorimeters that may need more non-trivial code.

**Comment by wdconinc** [comment]
> So, I suggest we keep this PR for tracker (and photomultiplier) hits only, and create a similar PR for calorimeters that may need more non-trivial code.

In particular, we want more than just the leading hit in
https://github.com/eic/EICrecon/blob/ec93c7427137ad78e80bb6de1a33616874de7b18/src/algorithms/calorimetry/CalorimeterHitDigi.cc#L190-L194
and that then requires that we decide on a weight strategy (`hit.getEnergy() / edep` probably).

File: src/algorithms/tracking/TrackerHitReconstruction.cc [hunk 2]
```diff
@@ -75,6 +75,7 @@ std::unique_ptr<edm4eic::TrackerHitCollection> TrackerHitReconstruction::process
             m_cfg.timeResolution,                            // in ns
             static_cast<float>(raw_hit.getCharge() / 1.0e6),   // Collected energy (GeV)
             0.0F);                                       // Error on the energy
+        rec_hits->setRawHit(raw_hit);
```
Comment (veprbl) [inline src/algorithms/tracking/TrackerHitReconstruction.cc:78]:
```suggestion
#if EDM4EIC_VERSION_MAJOR >= 7
                rec_hits->setRawHit(raw_hit);
#endif
```

File: src/algorithms/tracking/TrackerHitReconstruction.cc [hunk 3]
```diff
@@ -75,6 +75,9 @@ std::unique_ptr<edm4eic::TrackerHitCollection> TrackerHitReconstruction::process
             m_cfg.timeResolution,                            // in ns
             static_cast<float>(raw_hit.getCharge() / 1.0e6),   // Collected energy (GeV)
             0.0F);                                       // Error on the energy
+#if EDM4EIC_VERSION_MAJOR >= 7
+                rec_hits->setRawHit(raw_hit);
```
Comment (veprbl) [inline src/algorithms/tracking/TrackerHitReconstruction.cc:79]:
```suggestion
        rec_hits->setRawHit(raw_hit);
```

File: src/factories/calorimetry/CalorimeterHitDigi_factory.h [hunk 4]
```diff
@@ -18,6 +18,7 @@ class CalorimeterHitDigi_factory : public JOmniFactory<CalorimeterHitDigi_factor
 
     PodioInput<edm4hep::SimCalorimeterHit> m_hits_input {this};
     PodioOutput<edm4hep::RawCalorimeterHit> m_hits_output {this};
+    PodioOutput<edm4eic::MCRecoCalorimeterHitAssociation> m_assoc_output {this};
```
Comment (veprbl) [inline src/factories/calorimetry/CalorimeterHitDigi_factory.h:21]:
```suggestion
```

File: src/algorithms/calorimetry/CalorimeterHitReco.cc [hunk 5]
```diff
@@ -276,6 +277,9 @@ void CalorimeterHitReco::process(
             sid,
             lid,
             local_position);
+#if EDM4EIC_VERSION_MAJOR >= 7
+        recohit.setRawHit(rh);
+#endif
```
Comment (wdconinc) [inline src/algorithms/calorimetry/CalorimeterHitReco.cc:282]: @veprbl Do we care about unused variables when taking false `#if` branch? `[[maybe_unused]]`?

File: src/algorithms/calorimetry/CalorimeterHitReco.cc [hunk 6]
```diff
@@ -265,7 +266,7 @@ void CalorimeterHitReco::process(
         const decltype(edm4eic::CalorimeterHitData::local) local_position(pos.x() / dd4hep::mm, pos.y() / dd4hep::mm,
                                                                        pos.z() / dd4hep::mm);
 
-        recohits->create(
+        auto recohit = recohits->create(
```
Comment (veprbl) [inline src/algorithms/calorimetry/CalorimeterHitReco.cc:269]:
```suggestion
#if EDM4EIC_VERSION_MAJOR >= 7
        auto recohit =
#endif
        recohits->create(
```

File: src/algorithms/tracking/TrackerHitReconstruction.cc [hunk 7]
```diff
@@ -66,7 +67,7 @@ std::unique_ptr<edm4eic::TrackerHitCollection> TrackerHitReconstruction::process
         //      - XYZ segmentation: xx -> sigma_x, yy-> sigma_y, zz -> sigma_z, tt -> 0
         //    This is properly in line with how we get the local coordinates for the hit
         //    in the TrackerSourceLinker.
-        rec_hits->create(
+        auto rec_hit = rec_hits->create(
```
Comment (veprbl) [inline src/algorithms/tracking/TrackerHitReconstruction.cc:70]:
```suggestion
 #if EDM4EIC_VERSION_MAJOR >= 7
       auto rec_hit =
 #endif
       rec_hits->create(
```

File: src/algorithms/calorimetry/CalorimeterHitReco.cc (hunk 5 above)
Comment (veprbl) [reply src/algorithms/calorimetry/CalorimeterHitReco.cc:285]: done

**Review by veprbl — APPROVED** [review]

**Pull request merged**